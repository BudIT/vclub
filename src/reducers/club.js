import { combineReducers } from 'redux';
import { INITIALIZE } from 'vclub/constants/actionTypes';

import membersReducer from './club/members';
import dataReducer from './club/data';


const baseReducer = combineReducers({
  members: membersReducer,
  data: dataReducer,
});

export default function clubReducer(atom, action) {
  if (action.type === INITIALIZE) {
    const newAtom = action.payload;
    newAtom.data.authenticated = true;
    newAtom.data.authenticating = false;

    return newAtom;
  }

  return baseReducer(atom, action);
}
