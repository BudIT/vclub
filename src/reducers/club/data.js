import createReducer from 'vclub/redux/createReducer';

import { auth } from 'vclub/actions/core';


export default createReducer({
  [auth]: (atom) => ({
    ...atom,
    authenticating: true,
  }),
}, {
  authenticated: false,
  authenticating: false,
  currentRoom: 'chat',
});
