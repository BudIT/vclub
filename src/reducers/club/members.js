import createReducer from 'vclub/redux/createReducer';

import { memberEnter, memberLeave } from 'vclub/actions/core';


export default createReducer({
  [memberEnter]: (atom, member) => [member, ...atom],
  [memberLeave]: (atom, memberId) => atom.filter((member) => member.id !== memberId),
}, []);
