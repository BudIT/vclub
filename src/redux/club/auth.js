import actionCreator from 'borex-actions/actionCreator';
import commandCreator from 'borex-actions/commandCreator';
import withSideEffect from 'borex-actions/withSideEffect';
import createReducer from 'borex-reducers/createReducer';
import setIn from 'borex-reducers/setIn';


export const auth = actionCreator(
  withSideEffect((context, authData, remember = true) => {
    const { ioSocket, localStorage, Raven } = context;

    ioSocket.emit('auth', authData);

    Raven.setUserContext(authData);

    Raven.captureMessage('Authenticated', {
      level: 'info',
      logger: 'auth',
      extra: authData,
    });

    if (remember) {
      localStorage.setItem('storedAuth', JSON.stringify(authData));
    }
  }),
);

export const setRestoredData = actionCreator();

export const restoreAuth = commandCreator((context) => {
  const { dispatch, localStorage } = context;
  const authJSON = localStorage.getItem('storedAuth');

  if (authJSON) {
    dispatch(setRestoredData(JSON.parse(authJSON)));
  }
});

export const logOut = commandCreator(() => {
  document.location.reload();
});


export default createReducer(on => {
  on(auth, setIn('authenticating', () => true));
  on(setRestoredData, setIn('restored'));
});
