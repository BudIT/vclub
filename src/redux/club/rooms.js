import initialState from 'vclub/redux/initialClubState';

export const CHANGE_ROOM = 'club/room/change-room';

export function changeRoom(newRoomName) {
  return {
    type: CHANGE_ROOM,
    payload: newRoomName,
    meta: {
      remote: true,
      broadcast: true,
    },
  };
}

export default function reducer(state, action) {
  switch (action.type) {
    case CHANGE_ROOM:
      if (state.currentRoom === action.payload) {
        return state;
      }

      return {
        ...state,
        currentRoom: action.payload,
      };

    default:
      return state || initialState.rooms;
  }
}
