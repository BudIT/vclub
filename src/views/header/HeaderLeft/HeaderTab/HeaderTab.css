.tab
{
  color: #33363A;
  text-decoration: none;
  text-align: center;
  font-size: 22px;
  font-family: 'Rubik', helvetica, sans-serif;

  padding: 14px 16px;
  transition: 0.2s;
  display: inline-block;

  background-color: inherit;
  cursor: pointer;

  border-style: none;
  outline: none;
}

.tabCurrent {
  composes: tab;
  text-decoration: underline;
}

.tab:hover {
  background-color: #FFAA00;
  /*text-decoration: line-through;*/
}
