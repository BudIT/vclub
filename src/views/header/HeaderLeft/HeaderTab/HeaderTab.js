import React, { PropTypes } from 'react';

import compose from 'recompose/compose';
import { withHandlers } from 'recompose';

import style from './HeaderTab.css';

const enhance = compose(
  withHandlers({
    // array of handlers
    onClick: props => () =>
      props.dispatch(props.changeRoom(props.children)),
  })
);

// changeRoom
function HeaderTab(props) {
  const {
    children,
    currentRoom,

    onClick,
  } = props;

  return (
    <li>
      <button
        onClick={onClick}
        className={children === currentRoom ? style.tabCurrent : style.tab}
      >
        {children}
      </button>
    </li>
  );
}

HeaderTab.propTypes = {
  // should be only one child
  children: PropTypes.element.isRequired,
  currentRoom: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
};

export default enhance(HeaderTab);
