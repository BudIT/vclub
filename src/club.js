import React from 'react';
import io from 'socket.io-client';
import { render } from 'react-dom';
import { createStore, compose, applyMiddleware } from 'redux';

import { Provider } from 'react-redux';

import ClubLayout from 'vclub/views/clubLayout/ClubLayout';

import clientActionBroker from 'vclub/redux/middlewares/clientActionBroker';
import sideEffectProcessor from 'vclub/redux/middlewares/sideEffectProcessor';
import reducer from 'vclub/redux/clubReducer';
import initialState from 'vclub/redux/initialClubState';
import { restoreAuth } from 'vclub/redux/club/auth';


const ioSocket = io({ path: '/vclub-socket', forceNew: true });

const storeEnhancer = compose(
  applyMiddleware(
    clientActionBroker(ioSocket),
    sideEffectProcessor({ context: { ioSocket, localStorage } }),
  ),
  window.devToolsExtension ? window.devToolsExtension() : f => f
);

const store = createStore(reducer, initialState, storeEnhancer);

ioSocket.on('dispatch', action => store.dispatch(action));

store.dispatch(restoreAuth());

render((
  <Provider store={store}>
    <ClubLayout />
  </Provider>
), document.getElementById('AppRoot'));
