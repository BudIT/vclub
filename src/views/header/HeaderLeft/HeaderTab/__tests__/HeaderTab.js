/* eslint import/no-extraneous-dependencies: ["error", {"devDependencies": true}] */
/* eslint-env jest */
// jest.mock('react/lib/ReactDefaultInjection')

import { changeRoom } from 'vclub/redux/club/rooms';

import React from 'react';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import HeaderTab, { HeaderTabComponent } from '../HeaderTab';


const tabName = 'VIDEO';

const masterUser = {
  id: '1',
  name: 'User',
  master: true,
};

const justAMember = { ...masterUser, master: false };


test('<HeaderTab /> (current tab) renders correctly', () => {
  const wrapper = shallow(
    <HeaderTabComponent isCurrentTab onClick={jest.fn()}>
      {tabName}
    </HeaderTabComponent>
  );

  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

test('<HeaderTab /> (not current tab) renders correctly', () => {
  const wrapper = shallow(
    <HeaderTabComponent isCurrentTab={false} onClick={jest.fn()}>
      {tabName}
    </HeaderTabComponent>
  );

  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

test('HeaderTab dispatches action for master', () => {
  const dispatchSpy = jest.fn();
  const wrapper = mount(
    <HeaderTab isCurrentTab user={masterUser} dispatch={dispatchSpy}>
      {tabName}
    </HeaderTab>
  );

  const button = wrapper.find('button');
  button.simulate('click');
  expect(dispatchSpy).toHaveBeenCalledTimes(1);
  expect(dispatchSpy).toHaveBeenCalledWith(changeRoom(tabName));
});

test('HeaderTab ignores click for member', () => {
  const dispatchSpy = jest.fn();
  const wrapper = mount(
    <HeaderTab isCurrentTab user={justAMember} dispatch={dispatchSpy}>
      {tabName}
    </HeaderTab>
  );

  const button = wrapper.find('button');
  button.simulate('click');
  expect(dispatchSpy).not.toBeCalled();
});
