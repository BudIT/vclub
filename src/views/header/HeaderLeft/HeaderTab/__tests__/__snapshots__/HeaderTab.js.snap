exports[`test <HeaderTab /> (current tab) renders correctly 1`] = `
<li>
  <button
    className="tabCurrent"
    onClick={[Function]}>
    Видео
  </button>
</li>
`;

exports[`test <HeaderTab /> (not current tab) renders correctly 1`] = `
<li>
  <button
    className="tab"
    onClick={[Function]}>
    Видео
  </button>
</li>
`;
